### PR TITLE
2021 07 20 remote call multi

### PR DESCRIPTION
**Pull request by thedavidmeister**
implements multi-call for:

- `remote_call`
- `call`
- `get`
- `get_details`
- `get_links`
- `get_link_details`

ostensibly we only need `remote_call` for countersigning, but feedback from community is that having to sequentially loop over these common HDK functions is a pain point, so i enabled all of them to be async over a vector of inputs

### CODE REVIEW

**Comment by Connoropolous** [comment]
@thedavidmeister this is nice! will it return results in a predictable order that matches the request? And what about error catching?

**Comment by thedavidmeister** [comment]
@Connoropolous yeah same order in and out, error handling same as Rust native in that you'll get back the first error

if you're writing on the other end of one of these multi calls i recommend making the writes idempotent in case you need to retry

that was already the case but would be more likely with more calls

File: crates/holochain_cascade/src/lib.rs [hunk 1]
```diff
@@ -615,6 +636,17 @@ where
...
+    #[instrument(skip(self, options))]
+    /// Takes ownership of `self` to perform a `dht_get_links`.
+    /// Useful for async closures tht will drop a `&mut self` reference before the async completes.
+    pub async fn into_dht_get_links(
+        mut self,
+        key: WireLinkKey,
+        options: GetLinksOptions,
+    ) -> CascadeResult<Vec<Link>> {
+        self.dht_get_links(key, options).await
+    }
```
Comment (freesig) [inline crates/holochain_cascade/src/lib.rs:648]: I think this should be solved on the closure level by cloning the cascade not adding extra api to the cascade

File: crates/holochain_p2p/src/lib.rs [hunk 2]
```diff
@@ -122,6 +122,21 @@ pub struct HolochainP2pCell {
...
+        zome_name: ZomeName,
+        fn_name: FunctionName,
+        cap: Option<CapSecret>,
+        payload: ExternIO,
+    ) -> actor::HolochainP2pResult<SerializedBytes> {
+        self.call_remote(to_agent, zome_name, fn_name, cap, payload)
+            .await
+    }
+}
+
```
Comment (freesig) [inline crates/holochain_p2p/src/lib.rs:139]: Same

**Review by freesig — CHANGES_REQUESTED** [review]
This looks good. I think we shouldn't add `into` to the cascade and holochain_p2p api. You can just clone them in the closure.

File: crates/holochain_cascade/src/lib.rs (hunk 1 above)
Comment (thedavidmeister) [reply crates/holochain_cascade/src/lib.rs:648]: @freesig you would think so, but note the `Cascade` e.g. in `get.rs` is already owned so cloning won't do anything:

```rust
                        Cascade::from_workspace_network(
                            call_context.host_context.workspace(),
                            call_context.host_context.network().clone()
                        )
                        .into_dht_get(any_dht_hash, get_options)
```

so the error for the clone case looks like this:


```rust
error[E0515]: cannot return value referencing temporary value
  --> crates/holochain/src/core/ribosome/host_fn/get.rs:24:25
   |
24 |                            Cascade::from_workspace_network(
   |   _________________________^
   |  |_________________________|
   | ||
25 | ||                             call_context.host_context.workspace(),
26 | ||                             call_context.host_context.network().clone()
27 | ||                         )
28 | ||                         .clone()
   | ||________________________________- temporary value created here
29 | |                          .dht_get(any_dht_hash, get_options)
   | |____________________________________________________________^ returns a value referencing data owned by the current function
```

**Comment by thedavidmeister** [comment]
@freesig try now with the added `async` blocks

**Review by freesig — APPROVED** [review]

**Pull request merged**